File: python/mirage/core/langfuse/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.hierarchy.stat import make_stat
from mirage.core.langfuse.readdir import readdir
from mirage.core.langfuse.scope import detect_scope


def _session_extra(match: ScopeMatch) -> dict[str, str]:
    return {"session_id": match.slots["session_id"]}


def _prompt_extra(match: ScopeMatch) -> dict[str, str]:
    return {"prompt_name": match.slots["prompt_name"]}


def _dataset_extra(match: ScopeMatch) -> dict[str, str]:
    return {"dataset_name": match.slots["dataset_name"]}


stat = make_stat(
    detect_scope,
    readdir,
    extras={
        "session": _session_extra,
        "prompt": _prompt_extra,
        "dataset": _dataset_extra,
    },
)

File: typescript/packages/core/src/core/gmail/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GmailAccessor } from '../../accessor/gmail.ts'
import type { IndexEntry } from '../../cache/index/config.ts'
import type { PathSpec } from '../../types.ts'
import { FileStat, FileType } from '../../types.ts'
import { guessType } from '../../utils/filetype.ts'
import type { ScopeMatch } from '../hierarchy/scope.ts'
import { makeStat } from '../hierarchy/stat.ts'
import { readdir } from './readdir.ts'
import { detectScope } from './scope.ts'

function labelStat(_match: ScopeMatch, _path: PathSpec, entry: IndexEntry): FileStat {
  return new FileStat({
    name: entry.vfsName,
    type: FileType.DIRECTORY,
    extra: { label_id: entry.id },
  })
}

function dayStat(_match: ScopeMatch, _path: PathSpec, entry: IndexEntry): FileStat {
  return new FileStat({ name: entry.vfsName, type: FileType.DIRECTORY })
}

function messageStat(_match: ScopeMatch, _path: PathSpec, entry: IndexEntry): FileStat {
  return new FileStat({
    name: entry.vfsName,
    type: FileType.JSON,
    size: entry.size,
    extra: { message_id: entry.id, ...entry.extra },
  })
}

function attachmentDirStat(_match: ScopeMatch, _path: PathSpec, entry: IndexEntry): FileStat {
  return new FileStat({
    name: entry.vfsName,
    type: FileType.DIRECTORY,
    extra: { message_id: entry.id },
  })
}

function attachmentStat(_match: ScopeMatch, _path: PathSpec, entry: IndexEntry): FileStat {
  return new FileStat({
    name: entry.vfsName,
    type: guessType(entry.vfsName),
    size: entry.size,
    extra: { attachment_id: entry.id },
  })
}

export const stat = makeStat<GmailAccessor>(detectScope, readdir, {
  entryStats: {
    label: labelStat,
    day: dayStat,
    message: messageStat,
    attachment_dir: attachmentDirStat,
    attachment: attachmentStat,
  },
})
